staging/src/k8s.io/client-go/tools/cache/reflector_test.go

```diff
@@ -358,6 +360,79 @@ func TestReflectorListAndWatchWithErrors(t *testing.T) {
 	}
 }
 
+func TestReflectorListAndWatchInitConnBackoff(t *testing.T) {
+	maxBackoff := 50 * time.Millisecond
+	table := []struct {
+		numConnFails  int
+		expLowerBound time.Duration
+		expUpperBound time.Duration
+	}{
+		{5, 32 * time.Millisecond, 64 * time.Millisecond}, // case where maxBackoff is not hit, time should grow exponentially
+		{40, 35 * 2 * maxBackoff, 40 * 2 * maxBackoff},    // case where maxBoff is hit, backoff time should flatten
+
+	}
+	for _, test := range table {
+		t.Run(fmt.Sprintf("%d connection failures takes at least %d ms", test.numConnFails, 1<<test.numConnFails),
+			func(t *testing.T) {
+				stopCh := make(chan struct{})
+				connFails := test.numConnFails
+				fakeClock := clock.NewFakeClock(time.Unix(0, 0))
+				bm := wait.NewExponentialBackoffManager(time.Millisecond, maxBackoff, 100*time.Millisecond, 2.0, 1.0, fakeClock)
+				lw := &testLW{
+					WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+						iter := test.numConnFails - connFails
+						if connFails > 0 {
+							go func() {
+								// slight delay to avoid race and ensure the reflector is waiting on the timer before stepping its clock.
```

Comment:
Instead of this sleep I think you should poll the number of waiters and step as soon as it's non-zero. (In fact you could do that in a separate goroutine instead launching a new one here.)

A 3ms sleep is going to make this racy, I'm pretty sure.

Reply:
> A 3ms sleep is going to make this racy, I'm pretty sure.

💯

Reply:
Makes sense, I think that's what I've done now, let me know if I'm missing anything.